util/http/retry.go

```diff
@@ -27,13 +28,101 @@ import (
+	DefaultDialTimeoutMsec = 3000
+	// DefaultResponseHeaderTimeoutMsec is the default number of milliseconds before timeout while waiting for response header from a remote endpoint. See `TimeoutConfig.ResponseHeaderTimeout`.
+	DefaultResponseHeaderTimeoutMsec = 3000
+	// DefaultRequestTimeoutMsec is the default number of milliseconds that the entire request can take before timeout. See `TimeoutConfig.RequestTimeout`.
+	DefaultRequestTimeoutMsec = 30_000
+
+	// defaults based on a target total retry time of at least 5s. 30*((2^8)-1)>5000
+
+	// DefaultMaxRetries is the default number of retries that a retryable request will make. See `RetryConfig.MaxRetries`.
+	DefaultMaxRetries = 8
+	// DefaultMinWaitMsec is the default minimum number of milliseconds between attempts. See `RetryConfig.MinWait`.
+	DefaultMinWaitMsec = 30
+	// DefaultMaxWaitMsec is the default maxmimum number of millisends between attempts. See `RetryConfig.MaxWait`.
+	DefaultMaxWaitMsec = 300_000
 )
 
+// RetryConfig represents the settings for retries in a retryable http client.
+type RetryConfig struct {
+	// MaxRetries is the maximum number of retries before giving up on a retryable request.
+	// This does not include the initial request so the total number of attempts will be MaxRetries + 1.
+	MaxRetries int
+	// MinWait is the minimum wait time between attempts. The actual wait time is governed by the BackoffStrategy,
+	// but the wait time will never be shorter than this duration.
+	MinWait time.Duration
+	// MaxWait is the maximum wait time between attempts. The actual wait time is governed by the BackoffStrategy,
+	// but the wait time will never be longer than this duration.
+	MaxWait time.Duration
+}
+
+// TimeoutConfig represents the settings for timeout at various points in a request lifecycle in a retryable http client.
+type TimeoutConfig struct {
+	// DialTimeout is the maximum duration that connection can take before a request attempt is timed out.
+	DialTimeout time.Duration
+	// ResponseHeaderTimeout is the maximum duration waiting for response headers before a request attempt is timed out.
+	// This starts after the entire request body is uploaded to the remote endpoint and stops when the request headers
+	// are fully read. It does not include reading the body.
+	ResponseHeaderTimeout time.Duration
+	// RequestTimeout is the maximum duration before the entire request attempt is timed out. This starts when the
+	// client starts the connection attempt and ends when the entire response body is read.
+	RequestTimeout time.Duration
+}
+
+// RetryableClientConfig is the complete config for a retryable http client
+type RetryableClientConfig struct {
+	TimeoutConfig
+	RetryConfig
+}
+
+// NewRetriableClientConfig creates a new config with default values.
+// Users of `NewRetryableClient` should use this method to get a new
+// config and then overwrite values if desired.
+func NewRetriableClientConfig() RetryableClientConfig {
```

Comment:
nit: `NewRetriableClientConfig` -> `NewRetryableClientConfig`

Reply:
Also, not a part of your change, but could we change `fetching artifacts: Retrying request` to just `Retrying request` within `RetryStrategy`